sources/code/main/modules/extensions.ts

```diff
@@ -19,20 +19,22 @@ async function fetchOrRead(file:string, signal?:AbortSignal) {
  * 
  * **Experimental** – it is unknown if that would work properly for all themes.
  */
-async function parseImports(cssString: string, importCalls:string[], maxTries=5):Promise<string> {
+async function parseImports(cssString: string, importCalls: string[], maxTries=5):Promise<string> {
   const anyImport = /^@import .+?$/gm;
   if(!anyImport.test(cssString)) return cssString;
   const promises:Promise<string>[] = [];
+  const newImports = importCalls.slice();
```

Comment:
Why do you copy the array, actually? I think it could make sense when you pass it in loop, but given this was removed and things are more linear (each recursion calls another only one time, there's no tree-alike structure of recursion calls), I think it only increases the memory usage (since array needs to be recreated per recursion call; we might want to pass the reference to the array to keep it all in one data structure and only modify it accordingly.

Then, you could use `importCalls.push`, pass it instead `newImports` and do `importCalls.pop()` (or even `importCalls.length--` if you enjoy writing funny code 😆).

Reply:
Copying the array is not specific to getting relative paths to work, its to fix the bug with circular imports. Without this change, any retry would instantly fail because the file it is attempting to reference is already in `importCalls`.

Reply:
Yeah, but if you need to remove the last `file` on retry, why don't you `importCalls.pop()`? We can just revert the array this way, so when retrying, we have the same args as with the first try.

Reply:
> (...)  its to fix the bug with circular imports.

Yeah, this PR helped me way before to notice I've actually introduced this bug once implementing this feat. But copying the array doesn't resolve it, arrays aren't read-only here so we are free to push and pop stuff. This mechanism is quite comparable to how stack works as a data structure (we either loop over everything, put something on top or get something from the top within this function).

Reply:
Wouldn't we need to pop off every file that was parsed in the current iteration?

Reply:
I guess we could store a number of files to pop off, but that seems more complicated than just keeping another array. (This could be because I like functional programming and am a little averse to mutable objects 😆)

Reply:
> I guess we could store a number of files to pop off, but that seems more complicated than just keeping another array?

Actually, you made me think if there wouldn't be a better way to utilise the [`Promise.allSettled`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) in way we can handle each failure respectively to further decide on continuing the recursion. I guess tho modifying the code that much around this is a bit out of the scope for this PR... I might work on it in the future (or you might go with another PR if you want to toy with it yourself)...

Another smart way that would work the same you want in case of the cloning is to just store the previous value of `length` (I guess it's better to store a number than a whole array, especially with many imports and recursions going on) and literally revert to it, as mentioned before this is quite funny way to cut or extend the array to the specified size. And I think this would be more performant if you consider that GC allocates new memory space on each array clone and leaves it allocated even after exiting from the scope, although JS is an abstract enough language that you can never be too sure if it even truly clones the array, I guess it really matters on the implementation of the JS engine and if those are doing stuff in some kind of a lazy matter or whatever...

Reply:
Are we really that concerned about copying an array (essentially vector of pointers) in an almost entirely IO-bound task? I feel like at best we shave off an unmeasurably small amount of time at the cost of less readable code (imo).

But of course I'll change it if you think its a good idea.

Reply:
> (This could be because I like functional programming and am a little averse to mutable objects 😆)

I guess I've been also more around imperative programming with C for a while, hence I'm quite thinking about the memory aspects in more of a C dev way... In fact, I'd even got some plans to experiment with implementing a C types in Node.js (in way they can be used to link JS types with C by converting them to a proper `Buffer` that is similar to raw data allocated by C; this could be useful for any form of communication with C software, since I'm aware of many C programs either implementing IPC or some network protocol for communications is usually using C types like structs to share data).

Reply:
> Are we really that concerned about copying an array (essentially vector of pointers) in an almost entirely IO-bound task?

I guess I'm more aware of the memory it leaves (since on each clone, the amount of data stored is basically doubled). And from my experience, GC quite rarely drops a data even without any references left. This could mean for WebCord to gain a bit more of the memory leftovers that might eventually be GC-d, but in practise this might happen only in cases when memory's getting full or other scenarios that trigger the algorithm – since GC is quite complex topic and it isn't happening that much as you might think. This would be the case especially on more imports resolved in its memory.

I would still go for the method of storing `length`, e.g. in `lastState` variable instead of doing the same for the whole array, and then just `importCalls.length = lastState`. I don't see why that would hurt that much the code readability, it's quite concise change.

The change's still feels out-of-the-scope for this PR either way, so I've decided I will make myself a commit with what I think could resolve this. This way, this PR is oriented only on changes around supporting CSS relative imports and nothing else...

Reply:
Ok, so you just want me to leave it as is and then you'll commit your changes to this branch?

Reply:
Yeah, I've already commited here: https://github.com/SpacingBat3/WebCord/commit/92a6302bd985f23ea7807885113c6fb145268df1.

Reply:
I've also resolved the conflicts by removing your implementation of copying stuff. I've also went for `allSettled` since I've noticed some promises could be left un-awaited when messing with the same string (I guess this could cause some stuff like duplicate resolves to the same stuff). I've also went for the smarter way by only adding the files to `importCalls` only when everything else succeeds (by putting it as last `then`). Combining both of these and the fact I modify string in-place, not after resolving all promises but actually as a part of doing it.